### PR TITLE
Implement IEquatable to Memory and ReadOnlyMemory

### PR DESCRIPTION
**Pull request by Gnbrkm41**
Related: #32905

This PR includes the changes required to implement IEquatable<T> to Memory<T> and ReadOnlyMemory<T>, and tests that checks for correct behaviours of the interface.

### CODE REVIEW

**Comment by Gnbrkm41** [comment]
Note: VS went crazy and weren't doing any validation at all for me... Hope not, but there is a slim chance that I might have missed some silly mistakes.




```
Error	NETSDK1013	The TargetFramework value '' was not recognized. It may be misspelled. If not, then the TargetFrameworkIdentifier and/or TargetFrameworkVersion properties must be specified explicitly.	System.Memory (ref\System.Memory)	C:\Program Files\dotnet\sdk\3.0.100-preview3-010431\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.TargetFrameworkInference.targets	93	
Error	NETSDK1013	The TargetFramework value '' was not recognized. It may be misspelled. If not, then the TargetFrameworkIdentifier and/or TargetFrameworkVersion properties must be specified explicitly.	System.Memory.Tests	C:\Program Files\dotnet\sdk\3.0.100-preview3-010431\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.TargetFrameworkInference.targets	93	
Error	NETSDK1013	The TargetFramework value '' was not recognized. It may be misspelled. If not, then the TargetFrameworkIdentifier and/or TargetFrameworkVersion properties must be specified explicitly.	System.Memory (src\System.Memory)	C:\Program Files\dotnet\sdk\3.0.100-preview3-010431\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.TargetFrameworkInference.targets	93	

```
Might be nice if I can figure out why this is happening, happened on me when I was trying to work on VisualBasic.Core solution as well. But besides... please have a look for possible weird stuffs among the changes.

**Comment by ahsonkhan** [comment]
Depends on https://github.com/dotnet/coreclr/pull/23586 which is why CI is failing.

File: src/System.Memory/tests/ReadOnlyMemory/Equality.cs [hunk 1]
```diff
@@ -84,6 +84,58 @@ public static void EmptyMemoryNotUnified()
...
+
+            Assert.False(left.Equals(right));
+            Assert.False(right.Equals(left));
+        }
+
+        [Fact]
+        public static void EqualityThroughInterface_IncludesBase()
+        {
+            int[] array = { 42, 43, 44, 45, 46 };
+            IEquatable<ReadOnlyMemory<int>> left = new ReadOnlyMemory<int>(array, 1, 3);
```
Comment (ahsonkhan) [inline src/System.Memory/tests/ReadOnlyMemory/Equality.cs]: How about mismatched? Where one is `IEquatable` and the other is a regular `Memory<int>`?

File: src/System.Memory/tests/ReadOnlyMemory/Equality.cs [hunk 2]
```diff
@@ -84,6 +84,58 @@ public static void EmptyMemoryNotUnified()
...
+            IEquatable<ReadOnlyMemory<int>> right = new ReadOnlyMemory<int>(array, 2, 3);
+
+            Assert.True(left.Equals(right));
+            Assert.True(right.Equals(left));
+        }
+
+        [Fact]
+        public static void EqualityThroughInterface_Reflexivity()
+        {
+            int[] array = { 42, 43, 44, 45, 46 };
```
Comment (ahsonkhan) [inline src/System.Memory/tests/ReadOnlyMemory/Equality.cs]: While we are here, can we add some tests for T != int as well?

For instance T == char (from string), in here: https://github.com/dotnet/corefx/blob/d3eed78a3765aef0325384c61a4593570ae3622b/src/System.Memory/tests/ReadOnlyMemory/Strings.cs#L161

And T == string.

**Comment by ahsonkhan** [comment]
> Might be nice if I can figure out why this is happening

I am not sure why that's happening. Did you run build at the root of the repo first (from the command line)? Try that and then re-open the solution in VS. I don't see such errors in my instance of VS.

File: src/System.Memory/tests/ReadOnlyMemory/Equality.cs (hunk 1 above)
Comment (Gnbrkm41) [reply src/System.Memory/tests/ReadOnlyMemory/Equality.cs]: Sure, why not.

File: src/System.Memory/tests/ReadOnlyMemory/Equality.cs (hunk 2 above)
Comment (Gnbrkm41) [reply src/System.Memory/tests/ReadOnlyMemory/Equality.cs]: Would I do the same as the tests for `int`'s, but instead with `char` and `string`s?

**Comment by Gnbrkm41** [comment]
>Try that and then re-open the solution in VS.
Appears that VS restart fixed it. What a bug. 🤣

File: src/System.Memory/tests/ReadOnlyMemory/Equality.cs (hunk 2 above)
Comment (Gnbrkm41) [reply src/System.Memory/tests/ReadOnlyMemory/Equality.cs]: + Do you mean that we add tests that compare two Memory<string> (or Memory<char>) using IEquatable interfaces?

File: src/System.Memory/tests/Memory/Equality.cs [hunk 3]
```diff
@@ -120,6 +120,58 @@ public static void DefaultMemoryCanBeBoxed()
...
         }
 
+        [Fact]
+        public static void EqualityThroughInterface_True()
+        {
+            int[] array = { 42, 43, 44, 45, 46 };
+            IEquatable<Memory<int>> left = new Memory<int>(array, 2, 3);
+            IEquatable<Memory<int>> right = new Memory<int>(array, 2, 3);
+
+            Assert.True(left.Equals(right));
```
Comment (Gnbrkm41) [inline src/System.Memory/tests/Memory/Equality.cs]: Thinking about it again, this resolves to `object.Equals(object)` call. I think the test for type mismatch should go here instead, actually.

File: src/System.Memory/tests/ReadOnlyMemory/Equality.cs [hunk 4]
```diff
@@ -84,6 +84,58 @@ public static void EmptyMemoryNotUnified()
...
         }
 
+        [Fact]
+        public static void EqualityThroughInterface_True()
+        {
+            int[] array = { 42, 43, 44, 45, 46 };
+            IEquatable<ReadOnlyMemory<int>> left = new ReadOnlyMemory<int>(array, 2, 3);
+            IEquatable<ReadOnlyMemory<int>> right = new ReadOnlyMemory<int>(array, 2, 3);
+
+            Assert.True(left.Equals(right));
```
Comment (Gnbrkm41) [inline src/System.Memory/tests/ReadOnlyMemory/Equality.cs]: resolves to object.Equals(object) call. Again, I think the test for type mismatch should go here.

File: src/System.Memory/tests/ReadOnlyMemory/Equality.cs (hunk 1 above)
Comment (Gnbrkm41) [reply src/System.Memory/tests/ReadOnlyMemory/Equality.cs]: @ahsonkhan Please have a look at the review comments I left below. I think the tests for mismatched types should go in `EqualityThroughInterface_True` case.

**Comment by Gnbrkm41** [comment]
```
D:\a\1\s\.packages\microsoft.dotnet.apicompat\1.0.0-beta.19177.11\build\Microsoft.DotNet.ApiCompat.targets(72,5): error : CannotRemoveBaseTypeOrInterface : Type 'System.Memory<T>' does not implement interface 'System.IEquatable<System.Memory<T>>' in the implementation but it does in the contract. [D:\a\1\s\src\System.Runtime\src\System.Runtime.csproj]
D:\a\1\s\.packages\microsoft.dotnet.apicompat\1.0.0-beta.19177.11\build\Microsoft.DotNet.ApiCompat.targets(72,5): error : CannotRemoveBaseTypeOrInterface : Type 'System.ReadOnlyMemory<T>' does not implement interface 'System.IEquatable<System.ReadOnlyMemory<T>>' in the implementation but it does in the contract. [D:\a\1\s\src\System.Runtime\src\System.Runtime.csproj]
D:\a\1\s\.packages\microsoft.dotnet.apicompat\1.0.0-beta.19177.11\build\Microsoft.DotNet.ApiCompat.targets(85,5): error : ApiCompat failed for 'D:\a\1\s\artifacts\bin\System.Runtime\uapaot-Windows_NT-Release\System.Runtime.dll' [D:\a\1\s\src\System.Runtime\src\System.Runtime.csproj]
```

Shouldn't https://github.com/dotnet/corefx/pull/36501 fix this? it's been merged and they clearly implement these interfaces. (`src/Common/src/CoreLib/System/Memory.cs and ReadOnlyMemory.cs`)

File: src/System.Memory/tests/ReadOnlyMemory/Equality.cs (hunk 2 above)
Comment (Gnbrkm41) [reply src/System.Memory/tests/ReadOnlyMemory/Equality.cs]: Added the cases for `string` and `char`.

**Comment by ahsonkhan** [comment]
That just mirrorred the sources. We need the binary dependencies to be updated. I believe this PR will update it: https://github.com/dotnet/corefx/pull/36480

**Comment by Gnbrkm41** [comment]
The PR seems to have been closed due to merging conflicts; has there been other PRs that updated the binary dependencies?

**Comment by ahsonkhan** [comment]
Here's the next one: https://github.com/dotnet/corefx/pull/36596

**Comment by Gnbrkm41** [comment]
Can we try running tests now? Guess it's merged now (although it seems like Maestro had some hiccups).

**Comment by Gnbrkm41** [comment]
Among the checks that have just finished running, there were three failures, of which two (Packaging All Configurations x64_Debug, Windows UWP_NETNative_x86_Release) are ApiCompat failures:
```
D:\a\1\s\.packages\microsoft.dotnet.apicompat\1.0.0-beta.19205.6\build\Microsoft.DotNet.ApiCompat.targets(72,5): error : CannotRemoveBaseTypeOrInterface : Type 'System.Memory<T>' does not implement interface 'System.IEquatable<System.Memory<T>>' in the implementation but it does in the contract. [D:\a\1\s\src\System.Runtime\src\System.Runtime.csproj]
D:\a\1\s\.packages\microsoft.dotnet.apicompat\1.0.0-beta.19205.6\build\Microsoft.DotNet.ApiCompat.targets(72,5): error : CannotRemoveBaseTypeOrInterface : Type 'System.ReadOnlyMemory<T>' does not implement interface 'System.IEquatable<System.ReadOnlyMemory<T>>' in the implementation but it does in the contract. [D:\a\1\s\src\System.Runtime\src\System.Runtime.csproj]
D:\a\1\s\.packages\microsoft.dotnet.apicompat\1.0.0-beta.19205.6\build\Microsoft.DotNet.ApiCompat.targets(85,5): error : ApiCompat failed for 'D:\a\1\s\artifacts\bin\System.Runtime\uapaot-Windows_NT-Debug\System.Runtime.dll' [D:\a\1\s\src\System.Runtime\src\System.Runtime.csproj]
D:\a\1\s\.packages\microsoft.dotnet.apicompat\1.0.0-beta.19205.6\build\Microsoft.DotNet.ApiCompat.targets(72,5): error : CannotRemoveBaseTypeOrInterface : Type 'System.Memory<T>' does not implement interface 'System.IEquatable<System.Memory<T>>' in the implementation but it does in the contract. [D:\a\1\s\src\System.Memory\src\System.Memory.csproj]
D:\a\1\s\.packages\microsoft.dotnet.apicompat\1.0.0-beta.19205.6\build\Microsoft.DotNet.ApiCompat.targets(72,5): error : CannotRemoveBaseTypeOrInterface : Type 'System.ReadOnlyMemory<T>' does not implement interface 'System.IEquatable<System.ReadOnlyMemory<T>>' in the implementation but it does in the contract. [D:\a\1\s\src\System.Memory\src\System.Memory.csproj]
D:\a\1\s\.packages\microsoft.dotnet.apicompat\1.0.0-beta.19205.6\build\Microsoft.DotNet.ApiCompat.targets(85,5): error : ApiCompat failed for 'D:\a\1\s\artifacts\bin\System.Memory\uapaot-Windows_NT-Debug\System.Memory.dll' [D:\a\1\s\src\System.Memory\src\System.Memory.csproj]
```

And one (Windows x64_Debug) is test failure that is not related to the changes made in this PR: [NullReferenceException in System.Net.Http.Functional.Tests.GetAsync_IPv6LinkLocalAddressUri_Success.](https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fpull~2F36497~2Fmerge/test~2Ffunctional~2Fcli~2F/20190411.32/workItem/System.Net.Http.Functional.Tests)

**Comment by ahsonkhan** [comment]
I believe you have to add the exception to the System.Runtime API compat baseline, for uapaot.
https://github.com/dotnet/corefx/blob/ce62a086814fdfecea1d6ba6657c3fb88763e1d6/src/System.Runtime/src/ApiCompatBaseline.uapaot.txt#L1

So, just add the following to the file:
```text
CannotRemoveBaseTypeOrInterface : Type 'System.Memory<T>' does not implement interface 'System.IEquatable<System.Memory<T>>' in the implementation but it does in the contract.
CannotRemoveBaseTypeOrInterface : Type 'System.ReadOnlyMemory<T>' does not implement interface 'System.IEquatable<System.ReadOnlyMemory<T>>' in the implementation but it does in the contract. 
```

@ericstj, @safern, please correct me if I am mistaken.

**Comment by ahsonkhan** [comment]
Do you mind filing an issue for that (labeled `area-System.Net.Http`)?

System.Net.Http.Functional.Tests.SocketsHttpHandlerTest_HttpClientHandlerTest_Http2/GetAsync_IPv6LinkLocalAddressUri_Success
System.Net.Http.Functional.Tests.PlatformHandler_HttpClientHandlerTest/GetAsync_IPv6LinkLocalAddressUri_Success
System.Net.Http.Functional.Tests.PlatformHandler_HttpClientHandlerTest/GetAsync_IPv6LinkLocalAddressUri_Success
https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fpull~2F36497~2Fmerge/test~2Ffunctional~2Fcli~2F/20190411.32/workItem/System.Net.Http.Functional.Tests/analysis/xunit/System.Net.Http.Functional.Tests.PlatformHandler_HttpClientHandlerTest~2FGetAsync_IPv6LinkLocalAddressUri_Success
Windows.10.Amd64.ClientRS4.ES.Open-x64-Debug
```text
Unhandled Exception of Type System.NullReferenceException
Message :
System.NullReferenceException : Object reference not set to an instance of an object.
Stack Trace :
   at System.Net.Test.Common.LoopbackServer.CreateServerAsync(Func`2 funcAsync, Options options) in D:\a\1\s\src\Common\tests\System\Net\Http\LoopbackServer.cs:line 59
   at System.Net.Http.Functional.Tests.HttpClientHandlerTest.GetAsync_IPv6LinkLocalAddressUri_Success() in D:\a\1\s\src\System.Net.Http\tests\FunctionalTests\HttpClientHandlerTest.cs:line 301
--- End of stack trace from previous location where exception was thrown ---
```

**Comment by Gnbrkm41** [comment]
I think https://github.com/dotnet/corefx/pull/36497/commits/39a3f2a876debcc186cac0c569bd0fe0cf26192d should do from what I can see, but the checks for `Packaging All Configurations x64_Debug, Windows UWP_NETNative_x86_Release` are still failing due to the same reason (ApiCompat).

**Comment by ahsonkhan** [comment]
I am not sure why that is still failing. 

@ericstj, Any ideas?  Do we need to update the baseline under shims too?
https://github.com/dotnet/corefx/blob/a114f57d7bc10e15c93d818f2fd5be5c942d598f/src/shims/ApiCompatBaseline.uapaot.netstandard.txt#L78

I don't think it matters, but other files have a total issue count in the file that the S.Runtime one is missing.
https://github.com/dotnet/corefx/blob/ce62a086814fdfecea1d6ba6657c3fb88763e1d6/src/System.Runtime.Loader/src/ApiCompatBaseline.uapaot.txt#L31

**Comment by ericstj** [comment]
It looks to me like this is happening for System.Memory:
https://dev.azure.com/dnceng/public/_build/results?buildId=152393&view=logs&jobId=45bce49c-c830-5a83-092f-833753df204f&taskId=a227182a-5109-5e62-9f74-d28307a2d6d2&lineStart=1627&lineEnd=1629&colStart=1&colEnd=279


D:\a\1\s\.packages\microsoft.dotnet.apicompat\1.0.0-beta.19205.6\build\Microsoft.DotNet.ApiCompat.targets(72,5): error : CannotRemoveBaseTypeOrInterface : Type 'System.Memory<T>' does not implement interface 'System.IEquatable<System.Memory<T>>' in the implementation but it does in the contract. [D:\a\1\s\src\System.Memory\src\\**System.Memory.csproj**]

**Comment by ahsonkhan** [comment]
> System.Memory.csproj

Lol, totally missed that! I thought both errors were from `System.Runtime.csproj`. Good eye :)

So, the solution is to add the API compat baseline exception there as well.

The baseline file for System.Memory was removed earlier by @MichalStrehovsky (here - https://github.com/dotnet/corefx/pull/33956) since it wasn't necessary anymore. Add it back with the two entries.

https://github.com/dotnet/corefx/blob/c9e735d2569d9322453314fd3b0d015f62157113/src/System.Memory/src/ApiCompatBaseline.uapaot.txt

**Comment by Gnbrkm41** [comment]
Thought I could remove ones in in System.Runtime, apparently I need both 🦆

**Pull request merged**